### ISSUE TITLE
Ensure cookies are properly supported

### ISSUE THREAD

**Issue by cjstehno** [http-builder-ng/http-builder-ng#49]
It seems that the Java and Apache client implementations are handling cookies differently; this can be seen in the `Issue`-marked cookie tests in `HttpGetSpec` (and others). If the whole class suite is run, the second cookie test executed will contain unexpected cookies in the request (from the first cookie test). This only happens in the Java implementation, not Apache. I have made the cookie-based tests check with `contains` for now, but this should be investigated.

Its understandable that cookies would be stored across requests for the same client, that's what they do; however, we need to ensure that both clients exhibit the same behavior and that it is configurable if necessary.

While the issue seems to be in the Java client, it would be good to investigate both implementations and document their cookie support better as well as providing any additional configuration hooks.

**Comment by dwclark** [comment]
This happened when I fixed a locking issue this past week. Basically Apache Client was locking in multithreaded environments when the requests were too slow. The solution was to create a fresh cookie store every request when the requests are multi-threaded.

The rule going forward should probably be:

1) If the number of threads per client is 1, then allow cookie persistence across requests. The user has said that any access to the client will be synchronized by them, so we can assume that cookies shared across requests are fine.

2) Otherwise, requests should not shared cookies at all. Since any thread and any connection could service a request, it doesn't make sense to cookie persistence across requests. Apache Http Builder does this now. Java needs to be made consistent. Go ahead and assign it to me.

**Comment by cjstehno** [comment]
I think the cookie behavior in Apache is configurable and this seems important. Consider the case of wanting to emulate browser behavior. When you load web page, your client is using multiple requests to load page content across multiple threads with the same cookies. While this may not be the most common use of HttpBuilder, it's a valid one in my opinion.

I at least want to get some solid documentation of how we support cookies and make them configurable and consistent across implementations. 

I will tread carefully on this fixing this and be sure that you are in agreement on the PR before merging.

**Comment by dwclark** [comment]
So we also need an 3rd option, persist cookies for all requests. Apache won't support that out of the box. We will need a thread safe, non-locking version of this. Apache's BasicCookieStore has locking issues for this configuration...and they only have the BasicCookieStore. I'll have to think about making the collection/store that handles this option. 

I agree, the user of HttpBuilder should be able to configure this setting via an enum set at the HttpObjectConfig level.

**Comment by cjstehno** [comment]
A few notes, ideas I came up with related to the cookie stuff... I may be able to do some prototyping along these lines this weekend.

| Strategy | Description |
|---|---|
| NoCookies | Cookies are not stored or managed at all. |
| InMemoryPerInstanceCookies | Cookies are stored in memory and managed/shared across an instance of `HttpBuilder`. |
| PersistedPerInstanceCookies | Cookies are stored somewhere on the local filesystem and managed/shared across an instance of `HttpBuilder`. |
| InMemoryPerThreadCookies | Cookies are stored in memory and managed/shared only within the context of a single thread. |
| PersistedPerThreadCookies | Cookies are stored somewhere on the filesystem and managed/shared only within the context of a single thread. |
| *PerRequestCookies | ... managed/shared for a single request.|

I am not entirely sure that the per-thread approach is valid - need to understand more about the internal client threading (what is the `Executor` for - async I think). It might be better to omit the per-thread strategies and just provide a means of writing a custom strategy.

The per-request strategy seems more like the `NoCookies` strategy than a standalone strategy.

The important parts of a `CookieStrategy` are:

* version - the cookie specification version
* persistence
	* in-memory has no persistence
	* persisted is stored somewhere on the filesystem (default temp or specified location)

It seems that how these will be configured depends a bit on what will be required to configure the cookie support in the different client implementations.

## References
* https://tools.ietf.org/html/rfc6265

**Comment by dwclark** [comment]
Per thread won't work for async because we can't assume anything about the thread that is executing. We could in theory execute a request for host A on a thread then execute a request for host B on the same thread.

I'm mostly a fan of making the library work harder so that the users don't have as much to think about. I've been thinking more about this and I think we can get away with just forcing them to specify cookie version, persistent store location, and cookies enabled/disabled. The defaults would be 0, null, and true respectively. Meaning use Netscape style cookies, use in memory store, and enable cookies.

As for cookies themselves, I think we should switch internally to using the jdk HttpCookie and write custom CookieStore classes based on the persistence strategy. HttpCookie will parse cookie headers for us and we can switch the Set-Cookie header to parse to List<HttpCookie> instead of MapPairs. The default cookie method in HttpConfig would configure cookies with an infinite timeout if no timeout is specified (I think we do that now).

Every instance of HttpBuilder would then act as a separate browser by default; there will be a CookieStore per client which is shared among requests, but not with other client instances. This will be thread safe. We could also provide a single method at the object config level which would allow clients to share the same store. The signature would basically be:

`void shareCookiesWith(HttpBuilder client)`

allowing clients to share cookies in any possible configuration they want. We send cookies "when it makes sense" i.e. just like a browser.

We should also provide a property at the object config level that enables/disables cookies, it will be enabled by default. With this addition the cookie configuration at the object level becomes:

`int getCookieVersion()`
`void setCookieVersion(int)`
`File getCookieFile()`
`void setCookieFile(File)`
`void shareCookiesWith(HttpBuilder client)`

Hopefully with that we are covered. And for most cases, people should every have to bother with setting these because the defaults are sensible.

**Comment by cjstehno** [comment]
That sounds like a simple and solid approach. We can talk about it more next week - unless you are going to implement this over the weekend. ;-) I am going to "chase a squirrel" related to client/server testing this weekend, after I merge/publish the release - tell you more when we meetup.

**Comment by dwclark** [comment]
I did some more research at the airport waiting to come home. I think we actually don't have to write anything. The JDK comes with classes that do everything we need. We just need to use them for all clients. See:

[Cookie Handler](https://docs.oracle.com/javase/7/docs/api/java/net/CookieHandler.html)
[Cookie Manager](https://docs.oracle.com/javase/7/docs/api/java/net/CookieManager.html)
[Cookie Store](https://docs.oracle.com/javase/7/docs/api/java/net/CookieStore.html)
[Cookie Policy](https://docs.oracle.com/javase/7/docs/api/java/net/CookiePolicy.html)

I thought these were only useful for HttpURLConnection. It looks like we can use them generically. I'll verify that the implementation of CookieManager will work for us, but since it's already working fine for the Java client I'm guessing it will work fine for the others as well.

**Comment by dwclark** [comment]
I spent some time figuring out how these classes work. They are rather convoluted, but that's because the various cookie specs are also convoluted.

It turns out that their default InMemoryCookieStore has the same problems as Apache Cookie Store did, under heavy contention with slow requests it's easy to end up locking in the cookie store. I have implemented two cookie stores which should never lock, but still be thread safe. The file store version will also use the client thread pool to make file writes asynchronous. I also added a trick I started using at work of using a modulo lock to allow the number of locks in a system to be small-ish while still allowing for a high throughput. Once these are tested, I think it's just a matter of hooking in the CookieHandler/CookieManagers at the HttpBuilder base level and allowing the separate clients to use them.
